apps/documents/serializers.py

```diff
@@ -0,0 +1,143 @@
+        file_s3_path = f"{storage_credentials_id}/{wallet_id}/{vault_id}/{document.id}.{file_type_name}"
+
+        pre_signed_post = s3.generate_presigned_post(
+            Bucket=bucket,
+            Key=file_s3_path,
+            Fields={"acl": "public-read", "Content-Type": content_type},
+            Conditions=[
+                {"acl": "public-read"},
+                {"Content-Type": content_type}
+            ],
+            ExpiresIn=settings.S3_URL_LIFE
+        )
+
+        return {
+            'data': pre_signed_post,
+            'uri': f"s3://{bucket}/{file_s3_path}",
+            'document_id': document.id
+        }
+
+
+class DocumentSerializer(serializers.ModelSerializer):
+    document_type = EnumField(DocumentType, ints_as_names=True)
+    file_type = EnumField(FileType, ints_as_names=True)
+    upload_status = EnumField(UploadStatus, ints_as_names=True)
+
+    class Meta:
+        model = Document
+        exclude = ['shipment']
+        read_only_fields = ('owner_id', 'shipment', 'document_type', 'file_type', 'size', 's3_path')
+
+    class JSONAPIMeta:
+        included_resources = ['shipment']
+
+
+class DocumentUpdateSerializer(serializers.ModelSerializer):
+    upload_status = EnumField(UploadStatus, ints_as_names=True)
+
+    class Meta:
+        model = Document
+        fields = '__all__'
+        read_only_fields = ('owner_id', 'shipment', 'document_type', 'file_type', 'size', 's3_path')
+
+
+class DocumentListSerializer(serializers.ListSerializer):
+    """
+    This custom class allows us to add to the returned documents the one time s3 generated url
+    """
+    @property
+    def data(self):
+        # pass
+        data = super(DocumentListSerializer, self).data
+        super(DocumentListSerializer, self).data        # pylint: disable=expression-not-assigned
+
+        for item in data:
+            item.update({'url': Document.objects.get(id=item['id']).generate_presigned_url})
+
+        return ReturnList(data, serializer=self)
+
+
+class DocumentRetrieveSerializer(serializers.ModelSerializer):
+    document_type = EnumField(DocumentType, ints_as_names=True)
+    file_type = EnumField(FileType, ints_as_names=True)
+    upload_status = EnumField(UploadStatus, default=UploadStatus.COMPLETED, ints_as_names=True)
+
+    class Meta:
+        model = Document
+        exclude = ['s3_path', 'shipment']
+        read_only_fields = ('owner_id', 'shipment', 'document_type', 'file_type', 'size')
+
+    @classmethod
+    def many_init(cls, *args, **kwargs):
+        kwargs['child'] = cls()
+        return DocumentListSerializer(*args, **kwargs)
+
+    def validate_upload_status(self):
+        instance = self.instance
+        if instance.upload_status != UploadStatus.COMPLETED:
+            raise serializers.ValidationError('This document does not exist!')
+        return True
+
+    @property
+    def data(self):
+        """
+        We override the returned data to include the one time s3 generated url
```

Comment:
I don't think you need to do this if you add presigned_url to the included fields. https://stackoverflow.com/a/18396622/845745